be/src/util/buffered_stream.cpp

```diff
@@ -146,11 +146,29 @@ Status SharedBufferedInputStream::get_bytes(const uint8_t** buffer, size_t offse
     if ((sb.offset > offset) || (sb.offset + sb.size) < (offset + *nbytes)) {
         return Status::RuntimeError("bad construction of shared buffer");
     }
-    if (sb.buffer.capacity() == 0) {
-        sb.buffer.reserve(sb.size);
-        RETURN_IF_ERROR(_file->read_at_fully(sb.offset, sb.buffer.data(), sb.size));
+
+    int head_pad = 0;
+    if (_enable_block_cache) {
+        head_pad = sb.offset % config::block_cache_block_size;
+        if (sb.buffer.capacity() == 0) {
+            // we allocate enough room aligned with `block_cache_block_size`
+            // so cache_input_stream can use this buffer and do zero copy.
+            int tail_pad = (head_pad + sb.size) % config::block_cache_block_size;
+            if (tail_pad > 0) {
+                tail_pad = config::block_cache_block_size - tail_pad;
+            }
+            sb.buffer.reserve(sb.size + head_pad + tail_pad);
+            RETURN_IF_ERROR(
+                    _file->read_at_fully(sb.offset - head_pad, sb.buffer.data(), sb.size + head_pad + tail_pad));
```

Comment:
If the cache hit ratio is not high, it will cause some negative optimization because more data be read from remote?

Reply:
In any way we have to read same amount data from cache, so actually there is no penalty here. 

Let's say in old way, we want to read 1KB data, but actually we read 1MB data from cache? So there is no difference here that we pass 1KB or 1MB.

Reply:
> In any way we have to read same amount data from cache, so actually there is no penalty here.
> 
> Let's say in old way, we want to read 1KB data, but actually we read 1MB data from cache? So there is no difference here that we pass 1KB or 1MB.

OK